_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,322 @@
+date: 2023-09-15
+author: paul_mellor
+---
+
+Strimzi simplifies the deployment and management of Kafka clusters in a Kubernetes environment, ensuring a smooth and hassle-free experience.
+But the real fun begins once your cluster is up and running.
+That's when you can start thinking about how to interact with Kafka to stream data.
+In other words, Kafka doing the job it excels at.
+
+In this post, we'll dive into the essentials of developing a Kafka producer client that can send messages to a Strimzi-managed Kafka cluster.
+To illustrate these concepts, we'll walk through a basic example of a self-contained application that generates and produces messages to a specific Kafka topic.
+
+## Getting started with your producer client
+
+The first thing to consider when developing a producer client application is your preferred programming language. 
+
+After you've decided on that, as a bare minimum, your application must be able to connect to the Kafka cluster and use producers to send messages.
+
+Let's break down the essential steps to get started:
+
+1. Start by choosing the Kafka client library that speaks your choice of programming language. We use Java in this post, but you can use Python, .NET, and so on.
+
+2. Get the library through a package manager or by downloading it from the source.
+
+3. Import the classes and dependencies that your Kafka client will need into your code.
+
+4. Tell your client how to find and connect with your Kafka cluster, specifying an address and port, and, if required, security credentials.
+
+5. Create a producer object to subscribe to topics and produce messages to Kafka.
+
+    > A client can be a Kafka producer, and consumer, and admin. 
+
+6. Pay attention to error handling. It's tricky, but vitally important when connecting and communicating with Kafka.
+
+## Creating the Kafka producer client
+
+Let's get down to creating the producer client.
+Our brief is to create a client that operates asynchronously, and is equipped with basic error-handling capabilities. 
+The application implements the `Callback` interface, which provides a method for asynchronous handling of request completion in a background thread.
+
+### Prerequisites
+
+To be able to operate, the producer client needs the following in place:
+
+* A running kafka cluster 
+* A Kafka topic where it sends messages 
+
+We can specify the connection details and the name of the topic in our client configuration.
+
+> By default, cluster configuration has `auto.create.topics.enable=true`, which means a topic is automatically created on the first request.
+
+### Client constants
+
+Now, let's define some customizable constants that we'll also use with the producer.
+
+**BOOTSTRAP_SERVERS**
+
+With this constant, we can define a list of host/port pairs for establishing an initial connection to the Kafka cluster.
+For example, `localhost:9092` might be your starting point.
+In production, you can use a single load balancer, or a list of brokers.
```

Comment:
I have a bit of mixed feelings about this. I guess using `localhost:9092` makes things easier. But it also means it has no real relationship to Strimzi and it does not help users to connect this to Strimzi. So it is basically just a generic _Kafka producer_ guide.

Reply:
I extended the description.

> **BOOTSTRAP_SERVERS**
> 
> With this constant, we define the initial connection point to the Kafka cluster. 
> You can specify a list of host/port pairs to establish this connection.
> For a generic Kafka deployment, you might start with a value like `localhost:9092`. 
> However, when working with a Strimzi-managed Kafka cluster, it's important to use the bootstrap server address configured in your Strimzi deployment.
> 
> In Strimzi, the bootstrap server address is in the format `<cluster_name>-kafka-<listener-name>-bootstrap:<port>`. 
> This address allows an external Kafka producer to connect to the cluster through an advertised listener port (9092 or higher).
> In production environments, you can use a single load balancer, or a list of brokers in place of a single bootstrap server.

Reply:
Maybe you should point them instead to get the bootstrap address from the Kafka CR status? As that would handle the external listeners as well, which are needed if you don't include how to build this into Docker image and deploy in Kubernetes